File: 6.栈与队列/4.有效的括号/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        if (s.size() % 2) return 0;
        stack<char> stk;
        for (int i = 0; i < s.size(); i ++ )
        {
            if (s[i] == '(') stk.push(')');
            else if (s[i] == '[') stk.push(']');
            else if (s[i] == '{') stk.push('}');
            else if (stk.empty() || stk.top() != s[i]) return 0;
            else stk.pop();
        }
        return stk.empty();
    }
};
// @lc code=end
class Solution {
    public:
        bool isValid(string s) {
            if (s.size() % 2) return 0;
            stack<char> stk;
            int step = 0;
            while (step < s.size())
            {
                if (s[step] == '(' || s[step] == '[' || s[step] == '{')
                {
                    stk.push(s[step ++]);
                }
                else
                {
                    if (stk.empty()) return 0;
                    if ((s[step] == ')' && stk.top() == '(') || (s[step] == ']' && stk.top() == '[') || (s[step] == '}' && stk.top() == '{'))
                    {
                        stk.pop();
                        step ++;
                    }
                    else return 0;
                }
            }
            return stk.empty();
        }
    };

File: CodeTop/字节/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        for ch in s:
            if ch == '}':
                if not stk or stk[-1] != "{": return False
                else: stk.pop()
            elif ch == ")":
                if not stk or stk[-1] != "(": return False
                else: stk.pop()
            elif ch == "]":
                if not stk or stk[-1] != "[": return False
                else: stk.pop()
            else: stk.append(ch)
        return not stk
# @lc code=end

